### PR TITLE
#5337 TimelineEvent and DefaultScheduleEvent: provide builder

### PR DESCRIPTION
**Pull request by christophs78**
... and deprecate args constructor

### CODE REVIEW

**Comment by Rapster** [comment]
My bad I wasn't clear enough, I think event shoud be immutable object unless there are exceptional cases, and indeed I see we use setters sometimes... I'm not an expert in this component, so if you know it well, feel free to do some adjustments.

One thing for sure, is that we can't mix builder pattern and multi-constructors, for 8.0 we should deprecate all of the constructors with args.

By the way, if event has to be mutable maybe you can choose chain methods over build pattern: it's a  lot more simple to implement

**Comment by christophs78** [comment]
Ok, i´ll look into making the events immutable objects in a separate PR.

**Comment by christophs78** [comment]
@Rapster: Added one more commit - is this what you suggested/expected?

**Comment by christophs78** [comment]
Method-Chaining does not seem to be a good idea...
```
15-Nov-2019 21:51:58.186 SCHWERWIEGEND [http-nio-8080-exec-10] org.primefaces.application.exceptionhandler.PrimeExceptionHandler.logException Property [title] not writable on type [org.primefaces.model.DefaultScheduleEvent]
 javax.el.PropertyNotWritableException: Property [title] not writable on type [org.primefaces.model.DefaultScheduleEvent]
	at javax.el.BeanELResolver$BeanProperty.write(BeanELResolver.java:315)
	at javax.el.BeanELResolver$BeanProperty.access$100(BeanELResolver.java:275)
	at javax.el.BeanELResolver.setValue(BeanELResolver.java:120)
```

**Comment by tandraschko** [comment]
i would not change the setter. It's defined that setter are void, we should just add a additional Builder.

**Comment by christophs78** [comment]
I reverted the "method chaining" commit.

Also tried to update showcase and run into the following:

Old code:
```
DefaultScheduleEvent event = new DefaultScheduleEvent("Champions League Match", previousDay8Pm(), previousDay11Pm());
event.setDescription("Team A vs. Team B");
eventModel.addEvent(event);
```

New code:
```
DefaultScheduleEvent event = DefaultScheduleEvent.builder()
		.title("Champions League Match")
		.startDate(previousDay8Pm())
		.endDate(previousDay11Pm())
		.description("Team A vs. Team B")
		.build();
eventModel.addEvent(event);
```

This requires more code than before. I understand that we should deprecate/remove the args constructors from an architecture-standpoint. But from an usability and backward-compatibility-view - should we really deprecate/remove ALL the args constructors?

**Comment by tandraschko** [comment]
i think so
maybe we can also leave the all-args constructor?

File: src/main/java/org/primefaces/model/timeline/TimelineEvent.java [hunk 1]
```diff
@@ -146,6 +190,10 @@ public TimelineEvent(String id, T data, LocalDateTime startDate, LocalDateTime e
         this.styleClass = styleClass;
     }
 
+    /**
+     * @deprecated Use {@link TimelineEvent.Builder} instead.
+     */
+    @Deprecated
     public TimelineEvent(TimelineEvent<T> event) {
```
Comment (Rapster) [inline src/main/java/org/primefaces/model/timeline/TimelineEvent.java:197]: Add a `TimelineEvent#builder(TimelineEvent)` method to replace copy constructor

**Review by Rapster — COMMENTED** [review]
> Method-Chaining does not seem to be a good idea...
> 
> ```
> 15-Nov-2019 21:51:58.186 SCHWERWIEGEND [http-nio-8080-exec-10] org.primefaces.application.exceptionhandler.PrimeExceptionHandler.logException Property [title] not writable on type [org.primefaces.model.DefaultScheduleEvent]
>  javax.el.PropertyNotWritableException: Property [title] not writable on type [org.primefaces.model.DefaultScheduleEvent]
> 	at javax.el.BeanELResolver$BeanProperty.write(BeanELResolver.java:315)
> 	at javax.el.BeanELResolver$BeanProperty.access$100(BeanELResolver.java:275)
> 	at javax.el.BeanELResolver.setValue(BeanELResolver.java:120)
> ```

That's funny, the return type shouldn't be a problem here, did you modify the interface also? Anyway, let's not change the interface. Builder should be good enough

**Comment by christophs78** [comment]
Done. Showcase-PR: https://github.com/primefaces/showcase-facelift/pull/88

**Pull request merged**